### PR TITLE
EGM-certified photon IDs for 2016 data: cut based and MVA

### PR DESCRIPTION
**Pull request by ikrav**
This PR contains EGM certified cut-based and MVA photon IDs tuned on 80X samples. They are intended to be used on 2016 data. This new photon MVA uses CITK-based isolation calculation that caused several minor changes in EGM sequences that go outside of the new IDs themselves. This PR contains extra adjustments for using Tasks recently introduced in the framework rather than Sequences, although all old Sequences are kept as well for backward compatibility.

### CODE REVIEW

**Comment by rafaellopesdesa** [comment]
Thanks @ikrav . As a reference, this is the same PR as #18013 with all the commits for this particular PR cherry-picked over the most recent IB (as suggested by @kpedro88 )...

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test
to get some more fresh comparisons

File: RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.cc [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+  allMVAVars.varESEnOverRawE = superCluster->preshowerEnergy() / superCluster->rawEnergy();
+  allMVAVars.varESEffSigmaRR = effSigmaRR;
+  allMVAVars.varRho          = *rho; 
+  allMVAVars.varPhoIsoRaw    = (*phoPhotonIsolationMap)[phoRecoPtr];  
+  allMVAVars.varChIsoRaw     = (*phoChargedIsolationMap)[phoRecoPtr];
+  allMVAVars.varWorstChRaw   = (*phoWorstChargedIsolationMap)[phoRecoPtr];
+
+  //photon iso corrected:
+
+  double eA = effectiveAreas_.getEffectiveArea( abs(superCluster->eta()) );
```
Comment (slava77) [inline RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.cc]: std::abs? The plain abs is for integers
Looking at EffectiveAreas.cc, it already does std::abs inside.

File: RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.cc [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+  double phoIsoPtScalingCoeffVal = 0;
+  if( !isEndcapCategory( findCategory ( particle ) ) ) 
+    phoIsoPtScalingCoeffVal = phoIsoPtScalingCoeff_.at(0); // barrel case
+  else
+    phoIsoPtScalingCoeffVal =  phoIsoPtScalingCoeff_.at(1); //endcap case
+  double phoIsoCorr = (*phoPhotonIsolationMap)[phoRecoPtr] - eA*(*rho) - phoIsoPtScalingCoeffVal*phoRecoPtr->pt();
+
+  allMVAVars.varPhoIsoCorr = TMath::Max(phoIsoCorr, phoIsoCutoff_);
```
Comment (slava77) [inline RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.cc]: please use std::max instead (general comment against use of TMath:: if a native c++ method is available)

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 3]
```diff
@@ -230,7 +230,12 @@ def miniAOD_customizeCommon(process):
                   'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring15_50ns_V1_cff',
                   'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_25ns_nonTrig_V2p1_cff',
                   'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_50ns_nonTrig_V2p1_cff']
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: shouldn't the mvaPhotonID_Spring16_nonTrig_V1_cff  and cutBasedPhotonID_Spring16_V2p2_cff be added here to save them in miniAOD?

Comment (rafaellopesdesa) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: @slava77 @ikrav I would say not only added, but updated (why would we keep the 15 IDs?)

**Comment by ahinzmann** [comment]
When the main recommended photon ID changes, this config will also need to be updated accordingly:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/CommonTools/PileupAlgos/python/PhotonPuppi_cff.py#L8
https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/CommonTools/PileupAlgos/python/PhotonPuppi_cff.py#L21
https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/CommonTools/PileupAlgos/python/PhotonPuppi_cff.py#L28
Would there be a way to import "latest recommended" rather that change this file everytime photon ID is updated?

**Comment by ikrav** [comment]
Let me check this with the conveners. 
Ilya

**Comment by ikrav** [comment]
Slava, this is indeed a problem. Part of the problem is that this has been in use
in many measurements already for four months or so. We are discussing now what
we can do about it. 

Ilya

**Comment by slava77** [comment]
Hi Ilya,

I just wanted to check on the status of this PR.

**Comment by ikrav** [comment]
Hi Slava,

   I think we+conveners just converged on what to do with the two issues:
 1) the bug with abs
 2) the addition of new IDs into miniAOD making script
I should be able to commit the updates soon (a bit distracted by teaching and grading,
it is the finals week here at UNL).

Ilya

**Comment by slava77** [comment]
ping

**Comment by ikrav** [comment]
Just came to CERN, working on it today.
Ilya

**Comment by ikrav** [comment]
Closing this PR and opening a new one with the same content, plus changes requested in the exchanges above, plus rebased to the most recent 92X IB release. The new PR is #18645.

**Pull request closed without merging**